patch/1.21.4/ngx_lua-ssl_session_hostname.patch

```diff
@@ -0,0 +1,51 @@
+diff --git src/ngx_http_lua_ssl_certby.c src/ngx_http_lua_ssl_certby.c
+index b8e70dde..c3bfc790 100644
+--- src/ngx_http_lua_ssl_certby.c
++++ src/ngx_http_lua_ssl_certby.c
+@@ -870,6 +870,46 @@ ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
+ }
+ 
+ 
++int
++ngx_http_lua_ffi_ssl_session_hostname(ngx_http_request_t *r, char **name,
++    size_t *namelen, char **err)
++{
++    ngx_ssl_conn_t          *ssl_conn;
++
++    if (r->connection == NULL || r->connection->ssl == NULL) {
++        *err = "bad request";
++        return NGX_ERROR;
++    }
++
++    ssl_conn = r->connection->ssl->connection;
++    if (ssl_conn == NULL) {
++        *err = "bad ssl conn";
++        return NGX_ERROR;
++    }
++
++#if (defined(TLS1_3_VERSION)                                                   \
++     && !defined(LIBRESSL_VERSION_NUMBER) && !defined(OPENSSL_IS_BORINGSSL))
++
++    /*
++     * SSL_SESSION_get0_hostname() is only available in OpenSSL 1.1.1+,
++     * but servername being negotiated in every TLSv1.3 handshake
++     * is only returned in OpenSSL 1.1.1+ as well
++     */
++
++    *name = (char *) SSL_SESSION_get0_hostname(SSL_get0_session(ssl_conn));
```

Comment:
Why is there no error handling for `SSL_get0_session`, as in the stream below. 🤔

Reply:
The OpenSSL documentation mentions that `SSL_SESSION_get0_hostname` returns a pointer on the SESS, and if that pointer is passed directly to the FFI, could that lead to an unintended release of that pointer.  https://docs.openssl.org/3.4/man3/SSL_SESSION_get0_hostname/#description

Maybe we should copy it?

Reply:
- It seems that the stream part is not the same.

https://github.com/api7/apisix-nginx-module/pull/107/files#diff-b5ef9aa5a11ec9246cbd119b5fcbdb8d1a06459a3c36fbd8d94cb1bfc6306f80R35-R37
https://github.com/nginx/nginx/commit/46b9f5d389447b3b822ea71f5ac86ebc316c2975#diff-8d7f7ff147d0e7e0b3391ff99ab539916d3b547625f5f3d954a78d364b745f0bR606

- LuaJIT FFI's conversion is performed by passing the C pointer directly to Lua as pointer cdata.

I think we're better off avoiding passing references and instead passing values.

https://luajit.org/ext_ffi_semantics.html

![image](https://github.com/user-attachments/assets/c0854f89-5456-49bf-b656-268ec159ef61)